### PR TITLE
ENH: Python interfaces for Linear regression

### PR DESCRIPTION
**Pull request by KulikovNikita**
Refers to https://github.com/intel/scikit-learn-intelex/pull/1147

### CODE REVIEW

**Comment by KulikovNikita** [comment]
[http://intel-ci.intel.com/](http://intel-ci.intel.com/eda6ee89-a93c-f114-90c5-a4bf010d0e2e)

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda6f0a0-adf4-f193-82b6-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda6fffc-558e-f136-8f78-a4bf010d0e2e

File: sklearnex/dispatcher.py [hunk 1]
```diff
@@ -84,6 +87,12 @@ def get_patch_map():
         mapping['kneighborsregressor'] = mapping['knn_regressor']
         mapping['nearestneighbors'] = mapping['nearest_neighbors']
 
+        # LinearRegression
+        mapping.pop('linear')
+        mapping['linear'] = [[(linear_model_module,
+                               'LinearRegression',
+                               LinearRegression_sklearnex), None]]
```
Comment (Alexsandruss) [inline sklearnex/dispatcher.py:94]:
```suggestion
        # LinearRegression
        if daal_check_version((2023, 'P', 100)):
            mapping.pop('linear')
            mapping['linear'] = [[(linear_model_module,
                                   'LinearRegression',
                                   LinearRegression_sklearnex), None]]

```

File: onedal/linear_model/tests/test_linear_regression.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+def test_pickle(queue):
+    assert len(get_queues())
+    X, y = load_diabetes(return_X_y=True)
+    model = LinearRegression(fit_intercept=True)
+    model.fit(X, y, queue=queue)
+    expected = model.predict(X, queue=queue)
+
+    import pickle
+    dump = pickle.dumps(model)
+    model2 = pickle.loads(dump)
```
Comment (Alexsandruss) [inline onedal/linear_model/tests/test_linear_regression.py:51]: Avoid usage of pickle or made serialization more secure in another way

File: sklearnex/linear_model/tests/test_linear.py [hunk 3]
```diff
@@ -25,7 +25,8 @@ def test_sklearnex_import_liner():
     X = np.array([[1, 1], [1, 2], [2, 2], [2, 3]])
     y = np.dot(X, np.array([1, 2])) + 3
     linreg = LinearRegression().fit(X, y)
-    assert 'daal4py' in linreg.__module__
+    assert 'daal4py' in linreg.__module__ or 'sklearnex' in linreg.__module__
```
Comment (Alexsandruss) [inline sklearnex/linear_model/tests/test_linear.py:28]:
```suggestion
    if daal_check_version((2023, 'P', 100)):
        assert 'sklearnex' in linreg.__module__
    else:
        assert 'daal4py' in linreg.__module__

```

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: onedal/linear_model/tests/test_linear_regression.py (hunk 2 above)
Comment (KulikovNikita) [reply onedal/linear_model/tests/test_linear_regression.py:51]: It is test. There is no need in extra security measures in this case.

File: sklearnex/linear_model/tests/test_linear.py (hunk 3 above)
Comment (KulikovNikita) [reply sklearnex/linear_model/tests/test_linear.py]: Fixed

File: sklearnex/dispatcher.py (hunk 1 above)
Comment (KulikovNikita) [reply sklearnex/dispatcher.py]: Fixed

File: deselected_tests.yaml [hunk 4]
```diff
@@ -54,6 +54,10 @@ deselected_tests:
   # TODO: investigate copy failure of read-only buffer
   - linear_model/tests/test_coordinate_descent.py::test_read_only_buffer
 
+  # LinearRegression
```
Comment (Alexsandruss) [inline deselected_tests.yaml:57]: Reason of deselection?

File: onedal/dal.cpp [hunk 5]
```diff
@@ -33,8 +33,11 @@ ONEDAL_PY_INIT_MODULE(polynomial_kernel);
 ONEDAL_PY_INIT_MODULE(sigmoid_kernel);
 
 /* algorithms */
-ONEDAL_PY_INIT_MODULE(svm);
+#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 2023001ul
```
Comment (Alexsandruss) [inline onedal/dal.cpp:36]: Version is stored in INTEL_DAAL_VERSION which is defined in "services/library_version_info.h"
Also, it contains only digits, not `u1`

File: onedal/linear_model/tests/test_linear_regression.py (hunk 2 above)
Comment (Alexsandruss) [reply onedal/linear_model/tests/test_linear_regression.py:51]: CodeFactor check should be greened. I guess some exception for it should be specified here

File: setup.py [hunk 6]
```diff
@@ -455,7 +465,7 @@ def run(self):
         'onedal.primitives',
         'onedal.datatypes',
         'onedal.common'
-    ]),
+    ] + (['onedal.linear_model'] if str(ONEDAL_VERSION) >= "20230000" else [])),
```
Comment (Alexsandruss) [inline setup.py:468]: Versions should be compared as ints.

File: setup.py [hunk 7]
```diff
@@ -322,7 +325,7 @@ def build_oneapi_backend():
         includes=includes,
         eca=eca,
         ela=ela,
-        defines=[],
+        defines=[onedal_version_define],
```
Comment (Alexsandruss) [inline setup.py:328]: This define breaks consistency of version check in all python wrappers where version is received from version header

File: setup_sklearnex.py [hunk 8]
```diff
@@ -13,17 +13,27 @@
...
 sklearnex_version = (os.environ["SKLEARNEX_VERSION"] if "SKLEARNEX_VERSION" in os.environ
                      else time.strftime("%Y%m%d.%H%M%S"))
 
+dal_root = os.environ.get('DALROOT')
+
+if dal_root is None:
+    raise RuntimeError("Not set DALROOT variable")
+
+ONEDAL_MAJOR_BINARY_VERSION, ONEDAL_MINOR_BINARY_VERSION = get_onedal_version(
+    dal_root, 'binary')
```
Comment (Alexsandruss) [inline setup_sklearnex.py:34]: Binary versions are not used in this script

File: setup.py (hunk 7 above)
Comment (KulikovNikita) [reply setup.py:328]: Any other proposals?

File: setup_sklearnex.py (hunk 8 above)
Comment (KulikovNikita) [reply setup_sklearnex.py]: Fixed

File: deselected_tests.yaml (hunk 4 above)
Comment (KulikovNikita) [reply deselected_tests.yaml:57]: Difference is overflowed just a little bit

File: onedal/dal.cpp (hunk 5 above)
Comment (KulikovNikita) [reply onedal/dal.cpp]: Fixed

File: setup.py (hunk 6 above)
Comment (KulikovNikita) [reply setup.py]: fixed

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda8981d-b304-f117-9445-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda8a1ae-8e37-f10b-8735-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda8a5b6-33ee-f12a-8b11-a4bf010d0e2e

File: deselected_tests.yaml (hunk 4 above)
Comment (Alexsandruss) [reply deselected_tests.yaml:57]: I mean it should be added to file

File: setup.py (hunk 7 above)
Comment (KulikovNikita) [reply setup.py:328]: All these tweaks with versions are required for green CI just because 2023.1 have not been published yet. I will remove all this stuff ASAP after the release

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda92ec6-63ea-f17f-b012-a4bf010d0e2e

File: setup.py (hunk 7 above)
Comment (KulikovNikita) [reply setup.py]: Fixed

File: deselected_tests.yaml (hunk 4 above)
Comment (KulikovNikita) [reply deselected_tests.yaml]: Fixed

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda950e4-0dc9-f1bc-b78b-a4bf010d0e2e

File: onedal/linear_model/tests/test_linear_regression.py (hunk 2 above)
Comment (KulikovNikita) [reply onedal/linear_model/tests/test_linear_regression.py]: I believe it will be fixed

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda96743-6a03-f171-93eb-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda99990-0cec-f1fd-abce-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/eda9feda-8945-f1a1-950e-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edaa24ef-9028-f116-87ee-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edaa291f-6cab-f1e4-a5eb-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edaa2fd6-1554-f113-90c5-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
[http://intel-ci.intel.com/](http://intel-ci.intel.com/edab5d98-8711-f11b-a7fd-a4bf010d0e2e)

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edab7227-1cbf-f135-b370-a4bf010d0e2e

File: sklearnex/linear_model/linear.py [hunk 9]
```diff
@@ -13,6 +12,221 @@
...
+            self : object
+                Fitted Estimator.
+            """
+            if sklearn_check_version(
+                    '1.0') and not sklearn_check_version('1.2'):
+                self._normalize = _deprecate_normalize(
+                    self.normalize,
+                    default=False,
+                    estimator_name=self.__class__.__name__,
+                )
```
Comment (Alexsandruss) [inline sklearnex/linear_model/linear.py:113]: Move to _onedal_fit

File: sklearnex/linear_model/linear.py [hunk 10]
```diff
@@ -13,6 +12,221 @@
...
+        def _initialize_onedal_estimator(self):
+            onedal_params = {
+                'fit_intercept': self.fit_intercept,
+                'copy_X': self.copy_X}
+            self._onedal_estimator = onedal_LinearRegression(**onedal_params)
+
+        def _onedal_fit(self, X, y, sample_weight, queue=None):
+            assert sample_weight is None
+            if sklearn_check_version("1.2"):
+                self._validate_params()
```
Comment (Alexsandruss) [inline sklearnex/linear_model/linear.py:216]: Remove extra check

Comment (KulikovNikita) [reply sklearnex/linear_model/linear.py]: Fixed

File: sklearnex/linear_model/linear.py (hunk 9 above)
Comment (KulikovNikita) [reply sklearnex/linear_model/linear.py]: Fixed

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edab89e5-5670-f18f-b257-a4bf010d0e2e

File: deselected_tests.yaml [hunk 11]
```diff
@@ -60,6 +60,31 @@ deselected_tests:
...
+  - linear_model/tests/test_ransac.py::test_perfect_horizontal_line
+  - model_selection/tests/test_search.py::test_grid_search_pipeline_steps
+  - tests/test_common.py::test_estimators[RANSACRegressor(estimator=LinearRegression())-check_estimators_dtypes]
+  - tests/test_multiclass.py::test_ovr_multiclass
+  - tests/test_multiclass.py::test_ovr_multilabel
+  - tests/test_multiclass.py::test_ovr_binary
+
+  # Linear Regression does not support fancy data types
+  - tests/test_common.py::test_estimators[RandomForestClassifier()-check_requires_y_none]
+  - tests/test_common.py::test_estimators[RandomForestRegressor()-check_requires_y_none]
```
Comment (samir-nasibli) [inline deselected_tests.yaml:81]: Are we sure about this tests?

File: deselected_tests.yaml [hunk 12]
```diff
@@ -60,6 +60,31 @@ deselected_tests:
...
+  - tests/test_common.py::test_estimators[LinearRegression()-check_requires_y_none]
+
+  # No correct logging in algorithm
+  - ensemble/tests/test_voting.py::test_voting_verbose[estimator0]
```
Comment (samir-nasibli) [inline deselected_tests.yaml:85]: How this test depens on LinReg?

File: deselected_tests.yaml [hunk 13]
```diff
@@ -60,6 +60,31 @@ deselected_tests:
...
+  # Linear Regression - equation system is bad defined
+  - ensemble/tests/test_stacking.py::test_stacking_prefit[StackingRegressor-DummyRegressor-predict-final_estimator1-X1-y1]
+
+  # Linear Regression - no warning on passing sparse array
+  - linear_model/tests/test_base.py::test_linear_regression_pd_sparse_dataframe_warning
+
+  # Linear Regression - unsupported data type
+  - linear_model/tests/test_ransac.py::test_ransac_score
+  - linear_model/tests/test_ransac.py::test_ransac_predict
+  - linear_model/tests/test_ransac.py::test_perfect_horizontal_line
```
Comment (samir-nasibli) [inline deselected_tests.yaml:72]: Is it possible reuse daal4py primitives on sklearnex side for aligning this tests?

**Review by samir-nasibli — APPROVED** [review]
LGTM

File: onedal/datatypes/validation.py [hunk 14]
```diff
@@ -157,8 +174,14 @@ def _check_classification_targets(y):
...
-    valid = (isinstance(y, Sequence) or sp.isspmatrix(y) or hasattr(y, '__array__')) \
-        and not isinstance(y, str)
+    valid = (
+        isinstance(
+            y,
+            Sequence) or sp.isspmatrix(y) or hasattr(
+            y,
+            '__array__')) and not isinstance(
+                y,
+        str)
```
Comment (Alexsandruss) [inline onedal/datatypes/validation.py:184]: Unreadable

File: onedal/version.hpp [hunk 15]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (Alexsandruss) [inline onedal/version.hpp:1]: Empty line

File: onedal/version.hpp [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+*******************************************************************************/
+
+#pragma once
+
+#include "services/library_version_info.h"
+
+#define MAJOR_VERSION  __INTEL_DAAL__
+#define MINOR_VERSION  __INTEL_DAAL_MINOR__
+#define UPDATE_VERSION __INTEL_DAAL_UPDATE__
+#define ONEDAL_VERSION MAJOR_VERSION * 10000 + MINOR_VERSION * 100 + UPDATE_VERSION
```
Comment (Alexsandruss) [inline onedal/version.hpp:25]: It is equivalent of `INTEL_DAAL_VERSION` from "services/library_version_info.h". Replace it with original

File: deselected_tests.yaml [hunk 17]
```diff
@@ -60,6 +60,31 @@ deselected_tests:
...
+  - model_selection/tests/test_search.py::test_grid_search_pipeline_steps
+  - tests/test_common.py::test_estimators[RANSACRegressor(estimator=LinearRegression())-check_estimators_dtypes]
+  - tests/test_multiclass.py::test_ovr_multiclass
+  - tests/test_multiclass.py::test_ovr_multilabel
+  - tests/test_multiclass.py::test_ovr_binary
```
Comment (Alexsandruss) [inline deselected_tests.yaml:77]: These tests are too important to deselect them

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: onedal/version.hpp (hunk 16 above)
Comment (KulikovNikita) [reply onedal/version.hpp]: Fixed

File: onedal/version.hpp (hunk 15 above)
Comment (KulikovNikita) [reply onedal/version.hpp]: Fixed

File: onedal/datatypes/validation.py (hunk 14 above)
Comment (KulikovNikita) [reply onedal/datatypes/validation.py]: Fixed

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edabc2b4-6bc4-f1e0-8392-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edabcf7d-c92e-f176-93a9-a4bf010d0e2e

File: onedal/dal.cpp [hunk 18]
```diff
@@ -15,6 +15,7 @@
 *******************************************************************************/
 
 #include "onedal/common.hpp"
+#include "onedal/version.hpp"
```
Comment (Alexsandruss) [inline onedal/dal.cpp:18]: Why not use `services/library_version_info.h`?

**Comment by Alexsandruss** [comment]
Fix is needed for these fails in Precommit:

```FAILED tests/test_patching.py::test_patching[LinearRegression int8 fit]
FAILED tests/test_patching.py::test_patching[LinearRegression int8 predict]
FAILED tests/test_patching.py::test_patching[LinearRegression int8 score]
FAILED tests/test_patching.py::test_patching[LinearRegression int16 fit]
FAILED tests/test_patching.py::test_patching[LinearRegression int16 predict]
FAILED tests/test_patching.py::test_patching[LinearRegression int16 score]
FAILED tests/test_patching.py::test_patching[LinearRegression int32 fit]
FAILED tests/test_patching.py::test_patching[LinearRegression int32 predict]
FAILED tests/test_patching.py::test_patching[LinearRegression int32 score]
FAILED tests/test_patching.py::test_patching[LinearRegression int64 fit]
FAILED tests/test_patching.py::test_patching[LinearRegression int64 predict]
FAILED tests/test_patching.py::test_patching[LinearRegression int64 score]
FAILED tests/test_patching.py::test_patching[LinearRegression float16 fit]
FAILED tests/test_patching.py::test_patching[LinearRegression float16 predict]
FAILED tests/test_patching.py::test_patching[LinearRegression float16 score]
FAILED tests/test_patching.py::test_patching[LinearRegression uint8 fit]
FAILED tests/test_patching.py::test_patching[LinearRegression uint8 predict]
FAILED tests/test_patching.py::test_patching[LinearRegression uint8 score]
FAILED tests/test_patching.py::test_patching[LinearRegression uint16 fit]
FAILED tests/test_patching.py::test_patching[LinearRegression uint16 predict]
FAILED tests/test_patching.py::test_patching[LinearRegression uint16 score]
FAILED tests/test_patching.py::test_patching[LinearRegression uint64 fit]
FAILED tests/test_patching.py::test_patching[LinearRegression uint64 predict]
FAILED tests/test_patching.py::test_patching[LinearRegression uint64 score]
```

File: deselected_tests.yaml (hunk 11 above)
Comment (KulikovNikita) [reply deselected_tests.yaml]: Fixed.

File: deselected_tests.yaml (hunk 12 above)
Comment (KulikovNikita) [reply deselected_tests.yaml]: Probably as a way to estimate importance of each classifier

File: deselected_tests.yaml (hunk 13 above)
Comment (KulikovNikita) [reply deselected_tests.yaml]: Fixed

File: onedal/dal.cpp (hunk 18 above)
Comment (KulikovNikita) [reply onedal/dal.cpp:18]: For me these macros like __INTEL_DAAL_XYZ__ look weird and outdated. We developing a new part of the library and the old one receives little to no updates.

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edabe670-18e2-f17f-ab08-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edac51c1-b9a2-f148-a223-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edac5317-e2dc-f1e7-b0b3-a4bf010d0e2e

File: onedal/datatypes/validation.py [hunk 19]
```diff
@@ -279,7 +299,7 @@ def _check_n_features(self, X, reset):
             f"is expecting {self.n_features_in_} features as input.")
 
 
-def _num_features(X):
+def _num_features(X, fallback_1d=False):
```
Comment (samir-nasibli) [inline onedal/datatypes/validation.py:302]: This is changes in common primitives. Should be on separate PRs

Comment (KulikovNikita) [reply onedal/datatypes/validation.py:302]: Discussed. Decided that it is correct

File: deselected_tests.yaml (hunk 17 above)
Comment (KulikovNikita) [reply deselected_tests.yaml]: Fixed

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edac66d9-b7db-f15e-b012-a4bf010d0e2e

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/edac6c44-d357-f1b3-b257-a4bf010d0e2e

**Comment by Alexsandruss** [comment]
```
  File "build/__release_lnx/daal4py-3.8/onedal/datatypes/_data_conversion.py", line 29, in convert_one_to_table
    return _backend.to_table(arg)
ValueError: Found unsupported array type
=========================== short test summary info ============================
ERROR tests/test_patching.py
```
